### ISSUE TITLE
Fail depreciated k2 schema

### ISSUE THREAD

**Issue by davidewatson** [samsung-cnct/k2#262]
New:

Schemas are depreciated in whole or not at all. This is a result of the earlier design decision to only version the entire config file and not each individual stanza. This ticket is to add a check to fail the kraken.config role if the schema version claimed by the config file is marked as depreciated. See the comments for more detail.

Old:

# Task

Per #221 , before a property can be removed it must be depreciated first to give the user a chance to stop depending on it. Programmatically this is done by adding the property to the depreciated list within the k2 schema. For example [1] gives a hypothetical schema stanza. Since depreciated is an extension to JSON schema (see [PR #173](https://github.com/json-schema-org/json-schema-spec/pull/173) for the current discussion) no standard validator will print warnings. This ticket is to create an Ansible task which will print warnings as part of the kraken.config role.  

[1]

```yaml
"kubeConfig": {
"title": "A Kubernetes configuration",
"description": "The location and version of a container containing the Kubernetes hyperkube binary.",
"properties": {
"name": {
"default": "defaultKubeConfig",
"description": "Name of the Kubernetes configuration.",
"type": "string"
},
"hyperkubeLocation": {
"default": "gcr.io/google_containers/hyperkube",
"description": "Location of the Kubernetes container.",
"format": "uri",
"type": "string"
},
"version": {
"default": "v1.5.2",
"description": "Version of the hyperkube binary.",
"format": "symver",
"type": "string"
}
},
"required": [
"name"
],
"depreciated": [
"hyperkubeLocation"
],
"type": "object"
},
```

**Comment by coffeepac** [comment]
blocked by #258

**Comment by coffeepac** [comment]
@davidewatson  to fill in new details for this ticket

**Comment by davidewatson** [comment]
The logic to load and validate configuration files is in ansible/roles/kraken.config/tasks/main.yaml. There are four cases:

1) The configuration is unsupported because it was created before k2 had a schema (verified by checking for a version propery in the configuration file).
2) The configuration is unsupported because there are no schema files for the specified version (verified by checking if there is a schema file at schemas/config/{{ kraken_config_user.version }}/config.json).
3) The configuration is supported, but may not be soon (verified by checking if the corresponding schema has a depreciated field set. E.g.

```
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "id": "http://judkins.house/apis/k2/v1/config.json",

  "depreciated": true,

  "title": "k2 configuration file",
  "description": "Describes one or more Kubernetes clusters",
```

In this case we must check if the --force option was passed or not. If it was, then proceed, otherwise, fail and print a message explaining that the user must either update their configuration file, or passed the --force option (not recommended).

4) The configation is supported if it passes validatation.

**Comment by davidewatson** [comment]
This ticket is really only about step 3) above. Since we don't have any deprecated schema yet, the meat of this ticket is to determine how to unit test the ansible which implements depreciation.

**Comment by DStorck** [comment]
doesn't seem like the `deprecated` keyword has been merged into json-schema-spec yet.  https://github.com/json-schema-org/json-schema-spec/pull/173 
Is there another method we are thinking about to check if the schema is valid?

**Comment by davidewatson** [comment]
JSON schema does not support depreciation natively, probably because it is a policy-heavy concept. Hence "depreciated" is a property we invented. ;) As with the default additionalProperties=true property of JSON Schema, anything that is not recognized is ignored.

**Comment by DStorck** [comment]
Tiny detail - but we do need to make sure this field gets set as 'deprecated' and not 'depreciated', as is convention

**Comment by davidewatson** [comment]
Yes, it should be spelled correctly. ;)

**Comment by DStorck** [comment]
this is waiting on https://github.com/samsung-cnct/k2/pull/404 to be merged

**Comment by DStorck** [comment]
This will also need to include some documentation about how/when we deprecate and the potentials problems that could happen with a --force

**Comment by DStorck** [comment]
@dwat do we need a this to also be done in k2cli since their is also a --force option there?

**Comment by DStorck** [comment]
I am running tests now to see if this works in k2cli - if it needs some attention I will file another ticket for k2cli.

**Comment by DStorck** [comment]
k2cli is a go! All set. Just waiting for tests to pass.
